homeassistant/components/switch/rainmachine.py

```diff
@@ -1,208 +1,400 @@
-            self.device_mac.replace(':', ''), self.rainmachine_entity_id)
+    def zones(self) -> list:
+        """Return a list of active zones associated with this program."""
+        return [z for z in self._obj['wateringTimes'] if z['active']]
+
+    def _calculate_running_days(self) -> str:
+        """Calculate running days from an RM string ("0010001100")."""
+        freq_type = self._obj['frequency']['type']
+        freq_param = self._obj['frequency']['param']
+        if freq_type == 0:
+            return 'Daily'
+
+        if freq_type == 1:
+            return 'Every {0} Days'.format(freq_param)
+
+        if freq_type == 2:
+            return ', '.join([
+                DAYS[idx] for idx, val in enumerate(freq_param[2:-1][::-1])
+                if val == '1'
+            ])
+
+        if freq_type == 4:
+            return '{0} Days'.format('Odd' if freq_param == '1' else 'Even')
+
+        return None
 
     def turn_off(self, **kwargs) -> None:
         """Turn the program off."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._client.programs.stop(self.rainmachine_entity_id)
-        except exceptions.BrokenAPICall:
-            _LOGGER.error('programs.stop currently broken in remote API')
-        except exceptions.HTTPError as exc_info:
+            self.rainmachine.client.programs.stop(self._rainmachine_entity_id)
+            dispatcher_send(self.hass, PROGRAM_UPDATE_TOPIC)
+        except HTTPError as exc_info:
             _LOGGER.error('Unable to turn off program "%s"', self.unique_id)
             _LOGGER.debug(exc_info)
 
     def turn_on(self, **kwargs) -> None:
         """Turn the program on."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._client.programs.start(self.rainmachine_entity_id)
-        except exceptions.BrokenAPICall:
-            _LOGGER.error('programs.start currently broken in remote API')
-        except exceptions.HTTPError as exc_info:
+            self.rainmachine.client.programs.start(self._rainmachine_entity_id)
+            dispatcher_send(self.hass, PROGRAM_UPDATE_TOPIC)
+        except HTTPError as exc_info:
             _LOGGER.error('Unable to turn on program "%s"', self.unique_id)
             _LOGGER.debug(exc_info)
 
-    @Throttle(MIN_SCAN_TIME, MIN_SCAN_TIME_FORCED)
     def update(self) -> None:
         """Update info for the program."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._entity_json = self._client.programs.get(
-                self.rainmachine_entity_id)
-        except exceptions.HTTPError as exc_info:
+            self._obj = self.rainmachine.client.programs.get(
+                self._rainmachine_entity_id)
+
+            self._attrs.update({
```

Comment:
We should only store information in the attributes that is relevant in further explaining the state.

Reply:
Just because we have the data, doesn't mean that we should add it to our state machine.

Reply:
I've been a tad over-zealous. 😄 Question: what about attributes that don't relate directly to the state, but would still be useful in influencing the state? Example: I have an attribute for soil type; it'd be great to use this in my automations for the switch itself (e.g., "If the temperature goes above 80 degrees, stop watering zones with Clay Loam soil").

Reply:
That can stay. I guess that is similar to things like max_temp for climate devices.

Reply:
Thanks, @balloob. I've removed all frivolous attributes and all attributes that (a) aren't directly related to the state and (b) aren't useful in influencing the state through automation. Let me know your thoughts?